### PR TITLE
lexer: lowercase-only delimiters for quoted strings

### LINKED ISSUE #13626: New syntax error in string literals in comments in 5.3
This comment no longer parses with OCaml 5.3 but parsed with 5.2:

```ocaml
(** [x < min({x'|x'∊l})] *)
```

The error is:
```
File "./local/test.ml", line 1, characters 0-3:
1 | (** [x < min({x'|x'∊l})] *)
    ^^^
Error: This comment contains an unterminated string literal
File "./local/test.ml", line 1, characters 13-17:
1 | (** [x < min({x'|x'∊l})] *)
                 ^^^^
  String literal begins here
```

This might come from https://github.com/ocaml/ocaml/pull/12664 that slightly changed the lexing rules of string literals in comments for the purpose of checking the validity of them and accidentally allowing all `identchar` in them.

### PR DESCRIPTION
**Pull request by Octachron**
This PR fixes #13626 by limiting quoted string delimiters to alphabetical lowercase identifiers (`straße` but not `x1` nor `x'`).

### CODE REVIEW

**Comment by gasche** [comment]
Naive questions:
- The bug report appeared to also be about single-quote `'` being supported that previously was not, right? What is the impact of your PR on single quotes? Where are they hidden in the code?
- Is this conservative with respect to the lexing rules of 5.2? (How do I review this property?)

**Comment by Octachron** [comment]
1. The single quotes (and digits) are hidden in the definition of `ident_ext`:
```ocaml
let identchar = ['A'-'Z' 'a'-'z' '_' '\'' '0'-'9']
let identstart_ext = identstart | utf8
let identchar_ext = identchar | utf8
let identstart = lowercase | uppercase
let ident_ext = identstart_ext  identchar_ext*
```
compared to the new `delim_ext`:
```ocaml
let delim_ext = ( lowercase | uppercase | utf_8 ) *
```
(where `uppercase` is only added to benefits from the improved error messages).
2. You can compare with 5.2 with `git diff 5.2.. -- parsing/lexer.mll` . The difference with 5.2 is essentially that `lowercase*` is replaced by `delim_ext`. The absence of uppercase letters is then checked in the delimiter validation function, to either emit a syntax error or to continue lexing inside a comment.

**Comment by gasche** [comment]
Note to self: in the diff from 5.2, the `_ext` prefix generally means "extended" and that means "also with a restricted choice of extra utf8 characters", except for `validate_ext` where `ext` means "extension (point)".

File: parsing/lexer.mll [hunk 1]
```diff
@@ -456,6 +456,7 @@ let identchar = ['A'-'Z' 'a'-'z' '_' '\'' '0'-'9']
 let utf8 = ['\192'-'\255'] ['\128'-'\191']*
 let identstart_ext = identstart | utf8
 let identchar_ext = identchar | utf8
+let delim_ext = (lowercase | uppercase | utf8)*
```
Comment (gasche) [inline parsing/lexer.mll:459]: Maybe you could have a comment on the fact that `uppercase` is in fact rejected by the `validate_delim` check, we intentionally accept more to provide clearer / more specific error messages?

**Review by gasche — APPROVED** [review]
Thanks for the extra comments. I think that I now understand the change and agree that it is safe -- the set of quoted string delimiters of the PR is closer to the one of 5.2, so there are less risks of regressions similar to #13626.

File: parsing/lexer.mll (hunk 1 above)
Comment (Octachron) [reply parsing/lexer.mll:459]: I have added an error message to point out that we want the same error message for `{A|` and `{À|`.

**Comment by gasche** [comment]
Looks good to me. By default I assume that you will merge and cherry-pick in 5.3.

**Comment by Octachron** [comment]
Indeed.

**Comment by Octachron** [comment]
Cherry-picked on 5.3 as 66b424d9ad9 .

**Pull request merged**